contracts/src/Cluster.sol

```diff
@@ -1,330 +1,356 @@
-event MigrationDataPullCompleted(uint64 id, address operatorAddress, uint128 clusterVersion);
-event MigrationCompleted(uint64 id, address operatorAddress, uint128 clusterVersion);
-event MigrationAborted(uint64 id, uint128 clusterVersion);
-
-event MaintenanceStarted(address addr, uint128 clusterVersion);
-event MaintenanceFinished(uint128 clusterVersion);
-
-event NodeOperatorAdded(uint8 idx, NodeOperator operator, uint128 clusterVersion);
-event NodeOperatorUpdated(NodeOperator operator, uint128 clusterVersion);
-event NodeOperatorRemoved(address operatorAddress, uint128 clusterVersion);
-
-event SettingsUpdated(Settings newSettings, uint128 clusterVersion);
-
-event OwnershipTransferred(address newOwner, uint128 cluserVersion);
-
-contract Cluster {
-    using Bitmask for uint256;
-
-    // TODO: Should we just make all the fields public?
-
-    address owner;
-
-    NodeOperators nodeOperators;
+struct NodeOperator {
+    address addr;
+    bytes data;
+    bool maintenance;
+}
 
-    Keyspace[2] keyspaces;
-    uint64 keyspaceVersion;
+struct Keyspace {
+    address[] members;
+    uint8 replicationStrategy;
+}
 
-    Settings settings;
-    Migration migration;
-    Maintenance maintenance;
+contract Cluster is Ownable2Step {
+    using EnumerableSet for EnumerableSet.AddressSet;
+    using LibSort for *;
+
+    /*//////////////////////////////////////////////////////////////////////////
+                            CONSTANTS & ERRORS
+    //////////////////////////////////////////////////////////////////////////*/
+
+    // Maximum number of operators per cluster
+    uint256 constant MAX_OPERATORS = 256;
+    // Maximum number of operators per migration (uint8 limit for pullingCount)
+    uint256 constant MAX_MIGRATION_OPERATORS = 255;
+
+    error Unauthorized();
+    error MigrationInProgress();
+    error NoMigrationInProgress();
+    error InvalidOperator();
+    error OperatorNotFound();
+    error OperatorNotPulling();
+    error CallerNotOperator();
+    error OperatorExists();
+    error OperatorInKeyspace();
+    error SameKeyspace();
+    error WrongMigrationId();
+    error InvalidOperatorData();
+    error TooManyOperators();
+    error NotInitialized();
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                EVENTS
+    //////////////////////////////////////////////////////////////////////////*/
+
+    event ClusterInitialized(uint256 operatorCount);
+    event MigrationStarted(uint64 indexed id, address[] operators, uint8 replicationStrategy, uint64 keyspaceVersion, uint128 version);
+    event MigrationDataPullCompleted(uint64 indexed id, address indexed operator, uint128 version);
+    event MigrationCompleted(uint64 indexed id, address indexed operator, uint128 version);
+    event MigrationAborted(uint64 indexed id, uint128 version);
+
+    event NodeOperatorAdded(address indexed operator, uint128 version);
```

Comment:
The idea was to be able to event-source: get initial state and then update the state by applying events. 
For that all the necessary data needs to be within the events. 
Wdyt?